src/cmd/ksh93/sh/io.c

```diff
@@ -1881,6 +1881,25 @@ void sh_vexsave(Shell_t *shp,int fn,int fd,Spawnvex_f vexfun, void *arg)
 	}
 }
 
+int safefdnumber(Shell_t* shp, int sfd)
+{
+	int fd;
+
+	while(1)
```

Comment:
This looks like a fix from the Fedora patches you linked to earlier. But it is questionable on several grounds. Such as the assertion only  fd 3 is affected. Also, this `while(1)` block looks to be bogus since ksh is not multithreaded. So the enclosed `for()` loop should be invariant. And even if it was multithreaded this mechanism should be protected by a lock rather than trying to execute the enclosed code more than once. Finally, notice the `break;` at the bottom of this `while() { }` loop which means it is never run more than once.

I also don't understand the innermost `continue;` statement. You could remove it and the code would behave the same.

I don't doubt the the idea embodied in this patch is correct but the fix is highly suspect.

Reply:
You are right. I have made some fixes based on your comments.